narwhal/storage/src/lib.rs

```diff
@@ -2,9 +2,57 @@
 // SPDX-License-Identifier: Apache-2.0
 
 mod certificate_store;
+mod header_store;
 mod node_store;
+mod payload_store;
 mod proposer_store;
+mod vote_digest_store;
 
 pub use certificate_store::*;
+use dashmap::DashMap;
+pub use header_store::*;
 pub use node_store::*;
+pub use payload_store::*;
 pub use proposer_store::*;
+use std::collections::VecDeque;
+use std::hash::Hash;
+use std::sync::Arc;
+use tokio::sync::oneshot;
+use tokio::sync::oneshot::{Receiver, Sender};
+use tracing::warn;
+pub use vote_digest_store::*;
+
+// A simple pub/sub to notify subscribers when a value becomes available.
+#[derive(Clone)]
+struct NotifySubscribers<K: Eq + Hash + Clone, V: Clone> {
```

Comment:
Note that sui has `NotifyRead` struct that seem to provide equivalent API.
The main difference between this and NotifyRead is that 
(a) NotifyRead is cancellation safe and clean up memory when all subscribers drop (while NotifySubscribers only clears memory when notify is called)
(b) NotifyRead has a counter for num pending requests and therefore provides better visibility

Reply:
Thanks @andll for bringing `NotifyRead` to my attention - I believe it would make sense to use one then and `NotifyRead` seems indeed to be better in the above 2 points. I would like to use it but It is under the `authority` module in the `sui-core` crate - which I would like to avoid refer to in Narwhal - it seems to be too internal impl detail. What do you think If I would move this under the `mysten-common` ? Would it be the right place to be so could be shared with `Narwhal` as well?

Reply:
Yes, I agree that components shared between sui in narwhal should be in common and we should not reference sui-core in narwhal create (we probably even could not if we want to, because sui-core depends on narwhal)

Reply:
> we probably could not even if we want to, because sui-core depends on narwhal

Yes, we definitely do on the `consensus_handler` & `ConnectivityMonitor` stuff so highly possible we'll get into the circular dependency issue. Ok, I'll go ahead and merge this PR and open a follow up to move the `NotifyRead` and use it on the narwhal storage.